File: crates/lure-core/src/tool/schema.rs
//! 最小 JSON Schema 参数校验。
//!
//! 对齐上游 `Schema.validate_json_schema_value` 的核心子集：类型、required、enum、
//! 数值 min/max、字符串长度、对象/数组递归。complex 组合（oneOf/anyOf、pattern、
//! additionalProperties 深校验）留待需要时补齐。

use serde_json::Value;

/// 解析非 null 的类型名（`["string","null"]` -> `"string"`）。
fn resolve_type(schema: &Value) -> Option<(String, bool)> {
    match schema.get("type") {
        Some(Value::String(t)) => Some((t.clone(), false)),
        Some(Value::Array(types)) => {
            let nullable = types.iter().any(|t| t.as_str() == Some("null"));
            let name = types
                .iter()
                .find_map(|t| t.as_str().filter(|s| *s != "null"))
                .map(str::to_string);
            name.map(|n| (n, nullable))
        }
        _ => None,
    }
}

/// 校验 `value` 是否符合 schema 片段；返回错误消息（空表示通过）。
pub fn validate_value(value: &Value, schema: &Value, path: &str) -> Vec<String> {
    let label = if path.is_empty() { "parameter" } else { path };
    let mut errors = Vec::new();

    let Some((type_name, nullable)) = resolve_type(schema) else {
        return errors;
    };
    if nullable && value.is_null() {
        return errors;
    }

    let type_ok = match type_name.as_str() {
        "string" => value.is_string(),
        "integer" => value.is_i64() || value.is_u64(),
        "number" => value.is_number(),
        "boolean" => value.is_boolean(),
        "array" => value.is_array(),
        "object" => value.is_object(),
        _ => true,
    };
    if !type_ok {
        errors.push(format!("{label} should be {type_name}"));
        return errors;
    }

    if let Some(Value::Array(choices)) = schema.get("enum") {
        if !choices.contains(value) {
            errors.push(format!("{label} must be one of the allowed values"));
        }
    }

    match type_name.as_str() {
        "integer" | "number" => {
            if let Some(n) = value.as_f64() {
                if let Some(min) = schema.get("minimum").and_then(Value::as_f64) {
                    if n < min {
                        errors.push(format!("{label} must be >= {min}"));
                    }
                }
                if let Some(max) = schema.get("maximum").and_then(Value::as_f64) {
                    if n > max {
                        errors.push(format!("{label} must be <= {max}"));
                    }
                }
            }
        }
        "string" => {
            let len = value.as_str().map(|s| s.chars().count()).unwrap_or(0);
            if let Some(min) = schema.get("minLength").and_then(Value::as_u64) {
                if (len as u64) < min {
                    errors.push(format!("{label} must be at least {min} chars"));
                }
            }
            if let Some(max) = schema.get("maxLength").and_then(Value::as_u64) {
                if (len as u64) > max {
                    errors.push(format!("{label} must be at most {max} chars"));
                }
            }
        }
        "object" => {
            let obj = value.as_object().expect("已校验为 object");
            if let Some(Value::Array(required)) = schema.get("required") {
                for key in required.iter().filter_map(Value::as_str) {
                    if !obj.contains_key(key) {
                        errors.push(format!("missing required {}", subpath(path, key)));
                    }
                }
            }
            if let Some(Value::Object(props)) = schema.get("properties") {
                for (key, sub) in obj {
                    if let Some(prop_schema) = props.get(key) {
                        errors.extend(validate_value(sub, prop_schema, &subpath(path, key)));
                    }
                }
            }
        }
        "array" => {
            let items = value.as_array().expect("已校验为 array");
            if let Some(min) = schema.get("minItems").and_then(Value::as_u64) {
                if (items.len() as u64) < min {
                    errors.push(format!("{label} must have at least {min} items"));
                }
            }
            if let Some(max) = schema.get("maxItems").and_then(Value::as_u64) {
                if (items.len() as u64) > max {
                    errors.push(format!("{label} must have at most {max} items"));
                }
            }
            if let Some(item_schema) = schema.get("items") {
                for (i, item) in items.iter().enumerate() {
                    let item_path = if path.is_empty() {
                        format!("[{i}]")
                    } else {
                        format!("{path}[{i}]")
                    };
                    errors.extend(validate_value(item, item_schema, &item_path));
                }
            }
        }
        _ => {}
    }

    errors
}

fn subpath(path: &str, key: &str) -> String {
    if path.is_empty() {
        key.to_string()
    } else {
        format!("{path}.{key}")
    }
}

File: packages/core/src/tool/schema.ts
//! 最小 JSON Schema 参数校验（对齐 crates/lure-core/src/tool/schema.rs）。

import type { Json } from "../provider/types.js";
import { asObject } from "../provider/json.js";

function resolveType(schema: Record<string, Json>): [string, boolean] | undefined {
  const t = schema["type"];
  if (typeof t === "string") return [t, false];
  if (Array.isArray(t)) {
    const nullable = t.some((x) => x === "null");
    const name = t.find((x): x is string => typeof x === "string" && x !== "null");
    return name !== undefined ? [name, nullable] : undefined;
  }
  return undefined;
}

/// 校验 `value` 是否符合 schema 片段；返回错误消息（空表示通过）。
export function validateValue(value: Json, schema: Json, path: string): string[] {
  const label = path === "" ? "parameter" : path;
  const errors: string[] = [];
  const s = asObject(schema);
  if (s === undefined) return errors;
  const resolved = resolveType(s);
  if (resolved === undefined) return errors;
  const [typeName, nullable] = resolved;
  if (nullable && value === null) return errors;

  const typeOk =
    typeName === "string"
      ? typeof value === "string"
      : typeName === "integer"
        ? typeof value === "number" && Number.isInteger(value)
        : typeName === "number"
          ? typeof value === "number"
          : typeName === "boolean"
            ? typeof value === "boolean"
            : typeName === "array"
              ? Array.isArray(value)
              : typeName === "object"
                ? value !== null && typeof value === "object" && !Array.isArray(value)
                : true;
  if (!typeOk) {
    errors.push(`${label} should be ${typeName}`);
    return errors;
  }

  if (Array.isArray(s["enum"])) {
    if (!s["enum"].some((c) => jsonEqual(c, value))) {
      errors.push(`${label} must be one of the allowed values`);
    }
  }

  if (typeName === "integer" || typeName === "number") {
    if (typeof value === "number") {
      const min = typeof s["minimum"] === "number" ? s["minimum"] : undefined;
      if (min !== undefined && value < min) errors.push(`${label} must be >= ${min}`);
      const max = typeof s["maximum"] === "number" ? s["maximum"] : undefined;
      if (max !== undefined && value > max) errors.push(`${label} must be <= ${max}`);
    }
  } else if (typeName === "string") {
    const len = [...(typeof value === "string" ? value : "")].length;
    const min = typeof s["minLength"] === "number" ? s["minLength"] : undefined;
    if (min !== undefined && len < min) errors.push(`${label} must be at least ${min} chars`);
    const max = typeof s["maxLength"] === "number" ? s["maxLength"] : undefined;
    if (max !== undefined && len > max) errors.push(`${label} must be at most ${max} chars`);
  } else if (typeName === "object") {
    const obj = asObject(value) ?? {};
    if (Array.isArray(s["required"])) {
      for (const key of s["required"]) {
        if (typeof key === "string" && !(key in obj)) {
          errors.push(`missing required ${subpath(path, key)}`);
        }
      }
    }
    const props = asObject(s["properties"]);
    if (props !== undefined) {
      for (const [key, sub] of Object.entries(obj)) {
        const propSchema = props[key];
        if (propSchema !== undefined) {
          errors.push(...validateValue(sub, propSchema, subpath(path, key)));
        }
      }
    }
  } else if (typeName === "array") {
    const items = Array.isArray(value) ? value : [];
    const min = typeof s["minItems"] === "number" ? s["minItems"] : undefined;
    if (min !== undefined && items.length < min) errors.push(`${label} must have at least ${min} items`);
    const max = typeof s["maxItems"] === "number" ? s["maxItems"] : undefined;
    if (max !== undefined && items.length > max) errors.push(`${label} must have at most ${max} items`);
    const itemSchema = s["items"];
    if (itemSchema !== undefined) {
      for (let i = 0; i < items.length; i++) {
        const itemPath = path === "" ? `[${i}]` : `${path}[${i}]`;
        errors.push(...validateValue(items[i], itemSchema, itemPath));
      }
    }
  }

  return errors;
}

function subpath(path: string, key: string): string {
  return path === "" ? key : `${path}.${key}`;
}

function jsonEqual(a: Json, b: Json): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}
